components/image-builder/pkg/builder/docker.go

```diff
@@ -215,17 +215,36 @@ func (b *DockerBuilder) getAbsoluteImageRef(ctx context.Context, ref string, all
 	return b.Resolver.Resolve(ctx, ref, resolve.WithAuthentication(auth))
 }
 
-func findGitInit(init *csapi.WorkspaceInitializer) *csapi.GitInitializer {
-	if init.GetGit() != nil {
-		return init.GetGit()
+func buildManifest(manifest map[string]string, prefix string, init *csapi.WorkspaceInitializer) error {
+	if ci := init.GetComposite(); ci != nil {
+		for i, child := range ci.Initializer {
+			err := buildManifest(manifest, fmt.Sprintf("%scomp[%d].", prefix, i), child)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
 	}
-	for _, current := range init.GetComposite().GetInitializer() {
-		nestedGitInit := findGitInit(current)
-		if nestedGitInit != nil {
-			return nestedGitInit
+
+	if ci := init.GetGit(); ci != nil {
+		manifest[prefix+"Source"] = "git"
+		manifest[prefix+"CloneTarget"] = ci.CloneTaget
+		manifest[prefix+"RemoteURI"] = ci.RemoteUri
+		return nil
+	}
+
+	if ci := init.GetDownload(); ci != nil {
+		manifest[prefix+"Source"] = "download"
+
+		hash := sha256.New()
+		for _, f := range ci.Files {
+			_, _ = hash.Write([]byte(f.Digest))
 		}
+		manifest[prefix+"Files"] = fmt.Sprintf("%x", hash.Sum(nil))
```

Comment:
Wouldn't this trigger docker builds on every change in `.gitpod.Dockerfile` (or other files)? I believe this would be too aggressive.